app/models/concerns/taxonomix.rb

```diff
@@ -132,6 +132,59 @@ def scope_by_taxable_ids(scope)
         scope.where("#{table_name}.id IN (#{cached_ids.join(',')})")
       end
     end
+
+    def join_organizations
+      join_taxonomies(user_organizations, "tto")
+    end
+
+    def join_locations
+      join_taxonomies(user_locations, "ttl")
+    end
+
+    def join_taxonomies(taxonomies, join_name)
+      joins("LEFT JOIN #{TAXONOMY_JOIN_TABLE} AS #{join_name} "\
+        "ON #{table_name}.id = #{join_name}.taxable_id "\
+        "AND #{join_name}.taxable_type = '#{base_class.name}' "\
+        "AND #{join_name}.taxonomy_id IN (#{taxonomies.map(&:id).join(', ')})")
+    end
+
+    def taxonomy_join_scope
+      distinct.join_organizations.join_locations
+    end
+
+    def in_taxonomy_scope(&block)
+      scope = taxonomy_join_scope
+      clauses = []
+
+      unless !Organization.current && User.current.admin?
+        clauses << org_join_arel[:taxonomy_id].in(user_organizations)
+      end
+
+      unless !Location.current && User.current.admin?
+        clauses << loc_join_arel[:taxonomy_id].in(user_locations)
+      end
+
+      nodes = clauses.any? ? clauses.reduce(:and) : nil
```

Comment:
It's bit weird naming for it, could we come up with something more descriptive?
```suggestion
      statement = clauses.any? ? clauses.reduce(:and) : nil
```

Reply:
sure. naming is very hard sometimes

Reply:
Done!